src/int.rs

```diff
@@ -0,0 +1,469 @@
+        magnitude: Uint<LIMBS>,
+    ) -> CtOption<Self> {
+        CtOption::new(
+            Self(magnitude).negate_if_unsafe(is_negative).0,
+            !magnitude.ct_gt(&Int::MAX.0) | is_negative & magnitude.ct_eq(&Int::MIN.0),
+        )
+    }
+
+    /// Create an [`Int`] from an array of [`Word`]s (i.e. word-sized unsigned
+    /// integers).
+    #[inline]
+    pub const fn from_words(arr: [Word; LIMBS]) -> Self {
+        Self(Uint::from_words(arr))
+    }
+
+    /// Create an array of [`Word`]s (i.e. word-sized unsigned integers) from
+    /// an [`Int`].
+    #[inline]
+    pub const fn to_words(self) -> [Word; LIMBS] {
+        self.0.to_words()
+    }
+
+    /// Borrow the inner limbs as an array of [`Word`]s.
+    pub const fn as_words(&self) -> &[Word; LIMBS] {
+        self.0.as_words()
+    }
+
+    /// Borrow the inner limbs as a mutable array of [`Word`]s.
+    pub fn as_words_mut(&mut self) -> &mut [Word; LIMBS] {
+        self.0.as_words_mut()
+    }
+
+    /// Borrow the limbs of this [`Int`].
+    pub const fn as_limbs(&self) -> &[Limb; LIMBS] {
+        self.0.as_limbs()
+    }
+
+    /// Borrow the limbs of this [`Int`] mutably.
+    pub fn as_limbs_mut(&mut self) -> &mut [Limb; LIMBS] {
+        self.0.as_limbs_mut()
+    }
+
+    /// Convert this [`Int`] into its inner limbs.
+    pub const fn to_limbs(self) -> [Limb; LIMBS] {
+        self.0.to_limbs()
+    }
+
+    /// Convert to a [`NonZero<Int<LIMBS>>`].
+    ///
+    /// Returns some if the original value is non-zero, and false otherwise.
+    pub const fn to_nz(self) -> ConstCtOption<NonZero<Self>> {
+        ConstCtOption::new(NonZero(self), self.0.is_nonzero())
+    }
+
+    /// Convert to a [`Odd<Int<LIMBS>>`].
+    ///
+    /// Returns some if the original value is odd, and false otherwise.
+    pub const fn to_odd(self) -> ConstCtOption<Odd<Self>> {
+        ConstCtOption::new(Odd(self), self.0.is_odd())
+    }
+
+    /// Get the sign bit of this [`Int`] as `Choice`.
+    pub fn sign_bit(&self) -> Choice {
+        Choice::from((self.0.to_words()[LIMBS - 1] >> (Word::BITS - 1)) as u8)
+    }
+
+    /// View the data in this type as an [`Uint`] instead.
+    pub const fn as_uint(&self) -> Uint<LIMBS> {
+        self.0
+    }
+
+    /// Whether this [`Int`] is equal to `Self::MIN`.
+    pub fn is_minimal(&self) -> Choice {
+        Choice::from((self == &Self::MIN) as u8)
+    }
+
+    /// Whether this [`Int`] is equal to `Self::MAX`.
+    pub fn is_maximal(&self) -> Choice {
+        Choice::from((self == &Self::MAX) as u8)
+    }
+
+    /// Perform the two's complement "negate" operation on this [`Int`]:
+    /// map `self` to `(self ^ 1111...1111) + 0000...0001`
+    ///
+    /// Returns
+    /// - the result, as well as
+    /// - whether the addition overflowed (indicating `self` is zero).
+    ///
+    /// Warning: this operation is unsafe; when `self == Self::MIN`, the negation fails.
+    #[inline]
+    fn negate_unsafe(&self) -> (Self, Choice) {
```

Comment:
it does what is the same as `self.0.wrapping_neg()` does.

Reply:
Thanks for pointing this out! I adapted `wrapping_neg()` into `wrapping_negc()` to include the carry and still get the `is_zero` for free. In addition to having no duplicate code, this saves anywhere from 8% (`I256`) to 1% (`I4096`) in the multiplication benchmark :ok_hand:

Reply:
btw. why do you need is_zero in the first place?

Reply:
Well, I _needed_ it for `mul` and `div`; the fix you pointed out there makes knowing whether something is zero unnecessary (for now).